pelican/plugins/seo/seo_enhancer/html_enhancer/__init__.py

```diff
@@ -37,14 +37,19 @@ def __init__(self, file, output_path, path, open_graph=False, twitter_cards=Fals
             sitename=_settings.get("SITENAME"),
         )
 
-        # The canonical URL must be built with custom metadata if filled
-        # If not, fallback to the default URL name
+        # The canonical URL must be built with custom metadata if filled.
+        # This can come from either the canonical or save_as field.
+        # If both are absent, fallback to the default URL name
+        canonical = _metadata.get("canonical")
         save_as = _metadata.get("save_as")
-        _fileurl = save_as if save_as else getattr(file, "url")
 
-        self.canonical_link = CanonicalURLCreator(
-            siteurl=_settings.get("SITEURL"), fileurl=_fileurl,
-        )
+        if canonical:
+            self.canonical_link = CanonicalURLCreator(siteurl=canonical, fileurl=None,)
+        else:
+            _fileurl = save_as if save_as else getattr(file, "url")
+            self.canonical_link = CanonicalURLCreator(
+                siteurl=_settings.get("SITEURL"), fileurl=_fileurl,
+            )
```

Comment:
You probably can go for something like this:
```
        external_canonical = _metadata.get("external_canonical")
        if external_canonical:
            self.canonical_link = external_canonical
        else:
            save_as = _metadata.get("save_as")
            _fileurl = save_as if save_as else getattr(file, "url")

            self.canonical_link = CanonicalURLCreator(
                siteurl=_settings.get("SITEURL"), fileurl=_fileurl,
            )
```
We don't need to call `CanonicalURLCreator` in case of external canonical as the URL is already built.

Reply:
I tried this initially, but elsewhere it expects a class, not a string. It will give `AttributeError: 'str' object has no attribute 'create_url'`:

```
Test session starts (platform: darwin, Python 3.9.0, pytest 6.2.3, pytest-sugar 0.9.4)
rootdir: /Users/jorritfolmer/Documents/git/seo
plugins: pythonpath-0.7.3, cov-2.11.1, sugar-0.9.4
collecting ... 
 pelican/plugins/seo/tests/test_article_schema_creator.py ✓✓✓✓✓✓✓✓✓                                       17% █▋        
 pelican/plugins/seo/tests/test_breadcrumb_schema_creator.py ✓✓✓✓✓✓                                       28% ██▊       
 pelican/plugins/seo/tests/test_canonical_url_creator.py ✓✓                                               31% ███▎      
 pelican/plugins/seo/tests/test_content_title_analyzer.py ✓✓✓✓                                            39% ███▉      
 pelican/plugins/seo/tests/test_internal_link_analyzer.py ✓✓✓                                             44% ████▌     
 pelican/plugins/seo/tests/test_open_graph.py ✓✓✓✓✓✓                                                      56% █████▋    
 pelican/plugins/seo/tests/test_page_description_analyzer.py ✓✓✓                                          61% ██████▎   
 pelican/plugins/seo/tests/test_page_title_analyzer.py ✓✓✓                                                67% ██████▋   
 pelican/plugins/seo/tests/test_robots_file_creator.py ✓✓✓                                                72% ███████▎  
 pelican/plugins/seo/tests/test_seo_enhancer.py ✓✓✓✓✓✓                                                    83% ████████▍ 

――――――――――――――――――――――― TestSEOEnhancer.test_add_html_enhancements_to_file_with_custom_canonical ―――――――――――――――――――――――

self = <seo.tests.test_seo_enhancer.TestSEOEnhancer object at 0x106d256a0>
fake_article_custom_canonical = <seo.tests.conftest.FakeArticle object at 0x106d25430>
fake_seo_enhancer = <seo.seo_enhancer.SEOEnhancer object at 0x106d253d0>

    def test_add_html_enhancements_to_file_with_custom_canonical(
        self, fake_article_custom_canonical, fake_seo_enhancer
    ):
        """
        Test if add_html_to_file add SEO enhancements
        to HTML files by mocking open().
        In this case the canonical metadata field is set,
        so we expect the canonical url to be set to this.
        """
    
        path = "fake_output/fake_file.html"
>       fake_html_enhancements = fake_seo_enhancer.launch_html_enhancer(
            file=fake_article_custom_canonical, output_path="fake_output", path=path,
        )

pelican/plugins/seo/tests/test_seo_enhancer.py:197: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <seo.seo_enhancer.SEOEnhancer object at 0x106d253d0>
file = <seo.tests.conftest.FakeArticle object at 0x106d25430>, output_path = 'fake_output'
path = 'fake_output/fake_file.html', open_graph = False, twitter_cards = False

    def launch_html_enhancer(
        self, file, output_path, path, open_graph=False, twitter_cards=False
    ):
        """
        Call HTMLEnhancer for each article and page.
        Return a dict with all HTML enhancements.
        """
    
        html_enhancer = HTMLEnhancer(
            file=file,
            output_path=output_path,
            path=path,
            open_graph=open_graph,
            twitter_cards=twitter_cards,
        )
    
        html_enhancements = {
>           "canonical_tag": html_enhancer.canonical_link.create_url(),
            "breadcrumb_schema": html_enhancer.breadcrumb_schema.create_schema(),
        }
E       AttributeError: 'str' object has no attribute 'create_url'

pelican/plugins/seo/seo_enhancer/__init__.py:37: AttributeError

 pelican/plugins/seo/tests/test_seo_enhancer.py ⨯                                                         85% ████████▌ 

――――――――――――――――― TestSEOEnhancer.test_add_html_enhancements_to_file_with_custom_canonical_and_save_as ―――――――――――――――――

self = <seo.tests.test_seo_enhancer.TestSEOEnhancer object at 0x106d0b3a0>
fake_article_custom_canonical = <seo.tests.conftest.FakeArticle object at 0x106d0b610>
fake_seo_enhancer = <seo.seo_enhancer.SEOEnhancer object at 0x106d0b4c0>

    def test_add_html_enhancements_to_file_with_custom_canonical_and_save_as(
        self, fake_article_custom_canonical, fake_seo_enhancer
    ):
        """
        Test if add_html_to_file add SEO enhancements
        to HTML files by mocking open().
        In this case both canonical and save_as metadata fields are set,
        so we expect the canonical url to be prevail.
        """
    
        path = "fake_output/fake_file.html"
>       fake_html_enhancements = fake_seo_enhancer.launch_html_enhancer(
            file=fake_article_custom_canonical, output_path="fake_output", path=path,
        )

pelican/plugins/seo/tests/test_seo_enhancer.py:258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <seo.seo_enhancer.SEOEnhancer object at 0x106d0b4c0>
file = <seo.tests.conftest.FakeArticle object at 0x106d0b610>, output_path = 'fake_output'
path = 'fake_output/fake_file.html', open_graph = False, twitter_cards = False

    def launch_html_enhancer(
        self, file, output_path, path, open_graph=False, twitter_cards=False
    ):
        """
        Call HTMLEnhancer for each article and page.
        Return a dict with all HTML enhancements.
        """
    
        html_enhancer = HTMLEnhancer(
            file=file,
            output_path=output_path,
            path=path,
            open_graph=open_graph,
            twitter_cards=twitter_cards,
        )
    
        html_enhancements = {
>           "canonical_tag": html_enhancer.canonical_link.create_url(),
            "breadcrumb_schema": html_enhancer.breadcrumb_schema.create_schema(),
        }
E       AttributeError: 'str' object has no attribute 'create_url'

pelican/plugins/seo/seo_enhancer/__init__.py:37: AttributeError

 pelican/plugins/seo/tests/test_seo_enhancer.py ⨯✓✓                                                       91% █████████▏
 pelican/plugins/seo/tests/test_seo_report.py ✓✓✓                                                         96% █████████▋
 pelican/plugins/seo/tests/test_twitter_cards.py ✓✓                                                      100% ██████████
=============================================== short test summary info ================================================
FAILED pelican/plugins/seo/tests/test_seo_enhancer.py::TestSEOEnhancer::test_add_html_enhancements_to_file_with_custom_canonical
FAILED pelican/plugins/seo/tests/test_seo_enhancer.py::TestSEOEnhancer::test_add_html_enhancements_to_file_with_custom_canonical_and_save_as

Results (0.67s):
      52 passed
       2 failed
         - pelican/plugins/seo/tests/test_seo_enhancer.py:186 TestSEOEnhancer.test_add_html_enhancements_to_file_with_custom_canonical
         - pelican/plugins/seo/tests/test_seo_enhancer.py:247 TestSEOEnhancer.test_add_html_enhancements_to_file_with_custom_canonical_and_save_as
```

Reply:
Good catch, this will need a refactor to handle this case more properly. I'll create an issue. In the meantime, your solution is good to me.

Reply:
#34